make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+
+
+class NodeFactory:
+    """
+    Class responsible to build the different RecursiveTreeNodes depending on its alignment and other parameters.
+    """
+    @staticmethod
+    def build(alignment: MSA, prg_builder: "PrgBuilder", parent_node: Optional[RecursiveTreeNode] = None) -> RecursiveTreeNode:
+        """
+        Builds the correct node given the alignment and other parameters.
+        Note that the root has a None parent. For the root also, we need to first try to build a MultiIntervalNode and
+        then a MulticlusterNode,as we first want to split its intervals, and then cluster.
+        """
+        parent_node_is_leaf = isinstance(parent_node, LeafNode)
+        assert not parent_node_is_leaf, "Error on building a Recursive Tree node: parent node should never be a leaf"
+
+        min_match_length = prg_builder.min_match_length
+        nesting_level = NodeFactory._get_nesting_level(parent_node)
+
+        # if parent is multi interval, children should not be
+        skip_building_multi_interval_node = isinstance(parent_node, MultiIntervalNode)
+        if not skip_building_multi_interval_node:
+            all_intervals, match_intervals = NodeFactory._get_all_and_match_intervals(alignment, min_match_length)
+
+            has_a_single_interval = NodeFactory._infer_if_has_single_interval(all_intervals=all_intervals, match_intervals=match_intervals,
+                                                                nesting_level=nesting_level,
+                                                                max_nesting=prg_builder.max_nesting,
+                                                                alignment=alignment,
+                                                                min_match_length=min_match_length)
+
+            if not has_a_single_interval:
+                interval_subalignments = NodeFactory._partition_alignment_into_interval_subalignments(alignment,
+                                                                                                      all_intervals)
+                return MultiIntervalNode(nesting_level, alignment, parent_node, prg_builder, interval_subalignments)
+
+        # if parent is multi cluster, children should not be
+        skip_building_multi_cluster_node = isinstance(parent_node, MultiClusterNode)
+        if not skip_building_multi_cluster_node:
+            clustering_result = kmeans_cluster_seqs(alignment, min_match_length)
+            cluster_further = NodeFactory._infer_if_we_should_cluster_further(alignment, clustering_result)
+
+            if cluster_further:
+                cluster_subalignments = NodeFactory._get_subalignments_by_clustering(alignment, clustering_result)
+                return MultiClusterNode(nesting_level, alignment, parent_node, prg_builder,
+                                        cluster_subalignments)
+
+        # here, the node is a leaf
+        return LeafNode(nesting_level, alignment, parent_node, prg_builder)
```

Comment:
This does mostly right thing, but not in a clear way IMO

I propose this instead:
```
        min_match_length = prg_builder.min_match_length
        if parent_node is None:
            nesting_level = 0
        else:
            nesting_level = parent_node.nesting_level

        if isinstance(parent_node, MultiIntervalNode):
            clustering_result = kmeans_cluster_seqs(alignment, min_match_length)
            cluster_further = NodeFactory._infer_if_we_should_cluster_further(alignment, clustering_result)
            cluster_further &= (nesting_level + 1 < max_nesting)

            if cluster_further:
                cluster_subalignments = NodeFactory._get_subalignments_by_clustering(alignment, clustering_result)
                result = MultiClusterNode(nesting_level + 1, alignment, parent_node, prg_builder,
                                        cluster_subalignments)
            else:
                result = LeafNode(nesting_level, alignment, parent_node, prg_builder)
        elif isinstance(parent_node, MultiClusterNode):
            all_intervals, match_intervals = NodeFactory._get_all_and_match_intervals(alignment, min_match_length)

            has_a_single_interval = len(all_intervals) == 1
            if not has_a_single_interval:
                interval_subalignments = NodeFactory._partition_alignment_into_interval_subalignments(alignment,
                result = MultiIntervalNode(nesting_level, alignment, parent_node, prg_builder, interval_subalignments)
            else:
                result = LeafNode(nesting_level, alignment, parent_node, prg_builder)
        else:
            raise ValueError("Unsupported parent type for building")
        return result
```

I've made a lot of changes here, here's my rationale:
* Your rules for building MultiInterval or MultiClusterNodes were based on exclusion, but AFAIR if parent is MultiInterval, we attempt horizontal partitioning, and if parent is MultiClusterNode, we attempt vertical partitioning, and that's that. This code reflects this.
* nesting level: needs only to be incremented when building a MultiClusternode. Needs only to be tested for > max_nesting when building a MultiClusterNode
* function `_infer_if_has_single_interval` is weird. it doesnt infer if vertical partition has a single interval, it also tests for max nesting and size of alignment. max nesting should be tested elsewhere (see above point), and alignment size is captured by `IntervalPartitioner` I believe. thus i propose removing this function
* I  propose removing function `_get_nesting_level`, I think code above captures all cases

Reply:
The code is a replacement for the whole build() function

Reply:
> Your rules for building MultiInterval or MultiClusterNodes were based on exclusion, but AFAIR if parent is MultiInterval, we attempt horizontal partitioning, and if parent is MultiClusterNode, we attempt vertical partitioning, and that's that. This code reflects this.

That works except when the parent is `None`, i.e. the root. The root can be either a `MultiIntervalNode`, a `MultiClusterNode` or a `LeafNode`, and we need to try this order for the root, as we first try to cluster vertically and then horizontally (if it is a `MultiIntervalNode`, we build it. Otherwise, if it is a `MultiClusterNode` we build it. If not, then the root is a leaf). I actually think this exclusion-based rationale is not the most intuitive, but it works for all types of nodes being built. For us to add your changes, we will need a method to build the root, that tries to build it as the 3 types of nodes in that order, which is basically the current implementation

I think the proposed code bugs when we have a single nonmatch interval, e.g.:
```
>1
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
>2
AAAAAAAAAAAAAAAATAAAAAAAAAAAAAAAAA
>3
GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
>4
GGGGGGGGGGGGGGGGGGCGGGGGGGGGGGGGGG
```
In this case, we can't build a `MultiIntervalNode`, but we can build a `MultiClusterNode` and we should do it, instead of building a `LeafNode`

> nesting level: needs only to be incremented when building a MultiClusternode. Needs only to be tested for > max_nesting when building a MultiClusterNode

I like this suggestion. I have an example where the PRGs built with the current and the suggested code will be different, but I think is fine as they encode the same information... to investigate further, we can talk through this example.

> function _infer_if_has_single_interval is weird. it doesnt infer if vertical partition has a single interval, it also tests for max nesting and size of alignment. max nesting should be tested elsewhere (see above point), and alignment size is captured by IntervalPartitioner I believe. thus i propose removing this function

I agree with this suggestion, although we should take care when we have a single mismatch interval

> I propose removing function _get_nesting_level, I think code above captures all cases

For the suggested code, I agree

Reply:
* About your exclusion-based code: i see, I overlooked the root case, and now understand the logic. 
 Could we not then, do
```
if parent_node is None or isinstance(parent_node, MultiClusterNode):
   <same as above>
elif isinstance(parent_node, MultiIntervalNode):
  <same as above>
```
AFAICT at the first `build()`, for the root, we should simply try to vertical partition, and that code above will produce a `MultiIntervalNode` or a `Leaf`

* About the single nonmatch interval, yep I totally forgot about that. (would your tests have picked that up btw?)
So then can we do this?
```
has_a_single_match_interval = len(all_intervals) ==  len(match_intervals) == 1
```

Reply:
done